pkg/querier/engine/streaming/operator/aggregation.go

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package operator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/util/zeropool"
+)
+
+type Aggregation struct {
+	Inner    InstantVectorOperator
+	Start    time.Time
+	End      time.Time
+	Interval time.Duration
+	Grouping []string
+
+	remainingInnerSeriesToGroup []*group // One entry per series produced by Inner, value is the group for that series
+	remainingGroups             []*group // One entry per group, in the order we want to return them
+}
+
+type group struct {
+	labels               labels.Labels
+	remainingSeriesCount uint
+
+	// Sum and presence for each step.
+	sums    []float64
+	present []bool
```

Comment:
The reason I've done it this way is so that we can pool the float and bool slices with other operators (or even other kinds of aggregations, like `avg`) - if we had a `step` struct here, while I agree it would be safer, it would mean that we can't pool the slices with other operators or other aggregations.

Reply:
Ah okay, thanks.